storage/remote/otlptranslator/prometheus/normalize_name.go

```diff
@@ -130,9 +130,9 @@ func normalizeName(metric pmetric.Metric, namespace string, allowUTF8 bool) stri
 		translationFunc = func(r rune) bool { return !unicode.IsLetter(r) && !unicode.IsDigit(r) && r != ':' }
```

Comment:
I think treating other letters than alphanumerics and colon specially doesn't make any sense in UTF-8 mode, since Prometheus in UTF-8 mode allows _all_ valid UTF-8 characters. There must be a better way of doing this.

Reply:
I think we need to tokenize because we don't want to repeat units in the metric name.

Imagine `http.requests.seconds` with `s` as unit. We don't want to create a metrics called `http.requests.seconds.seconds`, so we check if the unit is already present in the token array. We could also use `strings.Contains()`, I'm just not sure which open is better 🤔

Reply:
I don't understand how it makes sense to tokenize metric names in UTF-8 mode based on the character set allowed in Prometheus legacy mode, though. I.e., treating characters that are not alphanumeric and not colons as separators doesn't make sense in UTF-8 mode. All characters should be treated the same, so long as they are valid UTF-8.

I think that instead of tokenizing metric names based on legacy mode rules, there should be some sort of solution for determining whether a suffix should be added or not. Maybe pattern matching? In any case, it's a completely separate problem from escaping characters, which is what non-UTF-8 mode hs to do.

Reply:
Fair enough!

I'm a bit confused about your acceptance criteria for this PR, though. 

> I tested normalizeName with the OTLP metric name "metric.+name+" and unit "°C", and the result comes out as "metric.name+°C" instead of "metric.name+_°C".

Did you mean it comes out as `metric.name_°C`? The second option, `metric.name+_°C` does look correct anyway ; that's a good catch and can easily be fixed! I'll update the PR with the fix.

> I don't understand how it makes sense to tokenize metric names in UTF-8 mode based on the character set allowed in Prometheus legacy mode, though. I.e., treating characters that are not alphanumeric and not colons as separators doesn't make sense in UTF-8 mode. All characters should be treated the same, so long as they are valid UTF-8 (if they're invalid, the metric should be rejected).

Well, the code was initially written that way, and I was looking for the smallest change possible to fix the bug we've discovered. I believe tokens are still useful here to help us avoid duplicated unit/type suffixes in the final results, but I can agree that the term "separators" does not make sense anymore.

A redesign of the codebase sounds useful, and that's one of the things we've been discussing in the OTel-Prometheus SIG calls, right? Should we work on the redesign before committing to the fix here? Or is it ok if I just fix the bug you described above?

Reply:
> Did you mean it comes out as metric.name_°C?

It's the opposite, it doesn't. Anyway, it's not an acceptance criteria, but a symptom of the UTF-8 mode doing the wrong thing, i.e. escaping valid characters.

> Well, the code was initially written that way, and I was looking for the smallest change possible to fix the bug we've discovered.

The thing is that this is the design flaw I originally pointed out, and have documented as https://github.com/prometheus/prometheus/issues/15534. The OTLP UTF-8 mode can't be considered as working well before it no longer escapes characters. It's simply the wrong approach to split the metric name into characters.

> A redesign of the codebase sounds useful, and that's one of the things we've been discussing in the OTel-Prometheus SIG calls, right? 

I'm not sure what you mean by re-design of the code base. All we need in this context is to fix the OTLP UTF-8 mode such that it no longer escapes characters. I believe it simply should not tokenize metric names, since it's going to be an unnecessary performance hit at the very least, and most likely a source of bugs.

Reply:
I will check with @jesusvazquez about how we might fix the UTF-8 mode.

Reply:
I can see how one can understand the current behavior as "we're escaping utf8 characters", but thats not really what's happening.

The problem is that the current code that tokenizes string is losing characters at the end. E.g for "metric.name+" we lose the "+". When adding type suffixes, we add _ at the beginning. It does look like we're escaping the last character, but it's a different thing.

Reply:
With that said, I agree with you that this approach of tokenizing words may be not be the way we do this translation

Reply:
In any case, tokenizing based on which characters are supported by legacy Prometheus naming mode makes no sense.

Reply:
I don't think https://github.com/prometheus/prometheus/issues/15534 can be considered fixed until the OTLP endpoint UTF-8 mode no longer tokenizes names. Figuring out whether to add suffixes should be done in a different way.

Reply:
Hey Owen(@ywwg)! This is an interesting thread Arve and Arthur are having about how utf8 incoming metric names should be handled in the OTLP endpoint.

I was wondering if you could chime in and share your thoughts since you have context on how utf8 metric names are being handled across the prometheus codebase for remote write and scraping etc.

My opinion after reading through is that we should have a consistent approach so users can expect a similar handling while also having present that this is OTel and past prometheus conventions might not apply (like metric suffixes etcetera)

Reply:
To clarify about metric suffixes, that is rather orthogonal to the discussion here, since it's instead controlled by the `AddMetricSuffixes` API parameter (not a user parameter in Prometheus, but e.g. in Grafana Mimir), and if I recall correctly the consensus at the Prometheus dev summit (Berlin?) was to keep metric suffixes even in UTF-8 mode. The discussion here is more about _how_ to implement suffix generation in UTF-8 mode.

Reply:
> Imagine `http.requests.seconds` with `s` as unit. We don't want to create a metrics called `http.requests.seconds.seconds`, so we check if the unit is already present in the token array. We could also use

In the Otel world, isn't this metric normally declared as "http.requests" with the unit of "seconds"? Why would an otel user put the unit in the metric name? Do we have evidence of metric names in otel including type/unit already or is this a pure hypothetical? And if this is not the otel standard, why are we bending over backward to fix it?

I ask because this normalization approach is starting to smell like "too much magic" -- we are trying to detect every possible way someone might construct a metric name that would look weird in prometheus, but I don't think it's the job of prometheus to try to clean up someone else's mess. 

In the case where the user is asking for prometheus normalization, we can provide guidance for how prometheus will translate names -- escape to underscores, append units, etc.  Users should follow those guidelines if they want the names to appear in a consistent way.
 
If a user chooses to ignore that with a silly name like "http..requests+seconds++" with a unit "seconds", then we follow the normalization convention we defined, and they'll get "http..requests+seconds++_seconds" in utf-8 mode or "http__requests_seconds___seconds" with escaped mode, or maybe with underscore-reduction,  "http_requests_seconds_seconds".  Garbage in / garbage out. 

For the original example in this thread,  "metric.+name+_°C" seems like the correct result -- simply the input name + underscore + unit, no other mutations.

Reply:
regarding the original bug ( #15362 ), this is an escaping collision situation (https://github.com/prometheus/client_golang/issues/1638) -- and two UTF-8 metrics that escape to the same prometheus metric name should be handled as part of that process.

Reply:
In the meantime, a consensus was reached offline w/ @ywwg and @ArthurSens. In UTF-8 mode, we will not try to detect pre-escaping suffixes (and hence not split up and re-join the metric name).